cpp/src/arrow/compute/kernels/codegen_internal.h

```diff
@@ -1169,6 +1169,19 @@ ArrayKernelExec GeneratePhysicalNumeric(detail::GetTypeId get_id) {
   }
 }
 
+template <template <typename... Args> class Generator, typename... Args>
+ArrayKernelExec GeneratePhysicalDecimalToPhysicalDecimal(detail::GetTypeId get_id) {
```

Comment:
This weird name is my doing, I suggested it based on [this discussion](https://github.com/apache/arrow/pull/11233#discussion_r716684464).

Reply:
Actually, `GenerateDecimal` already exists, so it seems like this shouldn't be necessary. Perhaps this can be refactored?

Reply:
`GenerateDecimal` doesn't have the right template parameters (it takes a `Type0` which is used to instantiate a `Generator`), not sure what to do about this. `GenerateDecimal` only has one use, so maybe we can inline it there and rename this to `GenerateDecimal`?

Reply:
You could reconcile the two implementations. Just adapt the other `GenerateDecimal` use case to pass the other template parameter differently (for example at the end rather than at the start).

Reply:
I'm not sure how I'd unify the two - changing `GenerateDecimal` would have an impact on `applicator::ScalarBinaryEqualTypes`, which affects other types besides decimal.

It looks similar to the situation with `GenerateVarBinaryBase`, which we use for comparisons and `GenerateTypeAgnosticVarBinaryBase`, which we use for `if_else`. One tries to be generic in its return type, the other one assumes covariance.

Reply:
I see, this is unfortunate. In this case, can you name it `GenerateDecimalToDecimal` and add a small docstring?

Reply:
Done.